apis/Google.Cloud.Firestore/Google.Cloud.Firestore/AggregateQuery.cs

```diff
@@ -0,0 +1,129 @@
+    /// </summary>
+    private const string CountAlias = "Count";
+
+    private readonly IReadOnlyList<Aggregation> _aggregations;
+
+    internal AggregateQuery(Query query) =>
+        _query = GaxPreconditions.CheckNotNull(query, nameof(query));
+
+    private AggregateQuery(Query query, IReadOnlyList<Aggregation> aggregations)
+    {
+        _query = GaxPreconditions.CheckNotNull(query, nameof(query));
+        _aggregations = aggregations;
+    }
+
+    internal AggregateQuery WithAggregation(Aggregation aggregation)
+    {
+        if (aggregation.Count != null)
+        {
+            aggregation.Alias = CountAlias;
+        }
+        var newAggregations = _aggregations == null ? new List<Aggregation>() : new List<Aggregation>(_aggregations);
+        newAggregations.Add(aggregation);
+        return new AggregateQuery(_query, newAggregations);
+    }
+
+    /// <summary>
+    /// Asynchronously takes a snapshot of the result after applying the aggregate functions on the query.
+    /// </summary>
+    /// <param name="cancellationToken">A cancellation token for the operation.</param>
+    /// <returns>A <see cref="AggregateQuerySnapshot"/> which contains results of the Aggregate functions.</returns>
+    public Task<AggregateQuerySnapshot> GetSnapshotAsync(CancellationToken cancellationToken = default) =>
+        GetSnapshotAsync(null, cancellationToken);
+
+    internal async Task<AggregateQuerySnapshot> GetSnapshotAsync(ByteString transactionId, CancellationToken cancellationToken)
+    {
+        IAsyncEnumerable<RunAggregationQueryResponse> responseStream = GetAggregationQueryResponseStreamAsync(transactionId, cancellationToken);
+        Timestamp? readTime = null;
+        Value count = null;
+        await responseStream.ForEachAsync(response =>
+        {
+            if (response.Result.AggregateFields != null)
+            {
+                response.Result.AggregateFields.TryGetValue(CountAlias, out count);
+            }
+            if (readTime == null && response.ReadTime != null)
+            {
+                readTime = Timestamp.FromProto(response.ReadTime);
+            }
+        }, cancellationToken).ConfigureAwait(false);
+
+        GaxPreconditions.CheckState(readTime != null, "The stream returned from RunAggregationQuery did not provide a read timestamp.");
+        var snapshot = new AggregateQuerySnapshot(this, readTime.Value, count?.IntegerValue);
+        return snapshot;
+    }
+
+    private IAsyncEnumerable<RunAggregationQueryResponse> GetAggregationQueryResponseStreamAsync(ByteString transactionId, CancellationToken cancellationToken)
+    {
+        RunAggregationQueryRequest request = new RunAggregationQueryRequest
+        {
+            Parent = _query.ParentPath,
+            StructuredAggregationQuery = ToStructuredAggregationQuery()
+        };
+        if (transactionId != null)

```

Comment:
You could just use `Transaction = transactionId ?? ByteString.Empty` in the object initializer, I suspect.
(And yes, this may be something that could be simplified in Query too... I haven't checked.)

Reply:
While using this we are encountering StatusCode="InvalidArgument", Detail="Invalid transaction." error. The error is occurring because the setter of transactionId also sets consistencySelectorCase_  to ConsistencySelectorOneofCase.Transaction and a transaction with an empty transaction id is an invalid transaction. So I think we should stick to the old way.

Reply:
Ah, right. Yup, that makes sense.